File: fortran/test_directSolvers.f90


! Test for the linear solvers
PROGRAM testSolvers

    write(*,*) 'Testing Cholesky : '
    call testChol

    write(*,*) 'Testing LU : '
    call testLU

END PROGRAM testSolvers


SUBROUTINE testChol()
    USE linearSolvers
    implicit NONE

    INTEGER n, i
    REAL , allocatable :: A(:,:) , b(:) , x(:) , L(:,:)

    ! Inputs
    n = 3
    allocate(A(n,n) , b(n) , L(n,n) , x(n))

    A(1,:) = (/  2.0 , -1.0 ,  0.0 /)
    A(2,:) = (/ -1.0 ,  2.0 , -1.0 /)
    A(3,:) = (/  0.0 , -1.0 ,  2.0 /)

    b = (/ 1 , 2 , 3/)

    ! Solve using LU
    call Chol(n, A, L)
    call solveChol(n , L , b , x)

    ! Output
    write(*,*) 'n = ' , n
    write(*,*) 'x = ' , x
    write(*,*) 'L = '
    DO i = 1 , n
        write(*,*) L(i,:)
    END DO

    deallocate(A, b, L, x)

END SUBROUTINE testChol


SUBROUTINE testLU()
    USE linearSolvers
    implicit NONE

    INTEGER n, i
    REAL , allocatable :: A(:,:) , b(:) , x(:)
    INTEGER , allocatable :: P(:)

    ! Inputs
    n = 3
    allocate(A(n,n) , b(n) , P(n) , x(n))

    A(1,:) = (/  2.0 , -1.0 ,  0.0 /)
    A(2,:) = (/ -1.0 ,  2.0 , -1.0 /)
    A(3,:) = (/  0.0 , -1.0 ,  2.0 /)

    b = (/ 1 , 2 , 3/)

    ! Solve using LU
    call LU(n, A, P)
    call solveLU(n , A , P , b , x)

    ! Output
    write(*,*) 'n = ' , n
    write(*,*) 'x = ' , x
    write(*,*) 'LU = '
    DO i = 1 , n
        write(*,*) A(i,:)
    END DO
    write(*,*) 'P = ', P

    deallocate(A, b, P, x)

END SUBROUTINE testLU

File: tests/test_directSolvers.cpp
using namespace std;

#include <iostream>
#include "solvers/directSolvers.h"
#include "solvers/conjugateGradient.h"


void testLU();
void testChol();
void testCG();

int main() {
    cout << "Testing Cholesky: " << endl;
    testChol();

    cout << endl << "Testing LU: " << endl;
    testLU();

    cout << endl << "Testing CG: " << endl;
    testCG();

    return 0;
}



void testChol() {
    int n = 3;
    double M[n*n], L[n*n], b[n], x[n];

    // Input
    M[0] = 2.0; M[1] = -1.0; M[2] = 0.0;
    M[3] = -1.0; M[4] = 2.0; M[5] = -1.0;
    M[6] = 0.0; M[7] = -1.0; M[8] = 2.0;

    b[0] = 1.0; b[1] = 2.0; b[2] = 3.0;


    // Solve system
    Chol(n, M, L);
    solveChol(n, L, b, x);


    // Output
    cout << "L = " << endl;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cout << L[i*n + j] << " , ";
        }
        cout << endl;
    }
    cout << endl;

    cout << "x = " << endl;
    for (int i = 0; i < n; ++i) {
        cout << x[i] << " , ";
    }
    cout << endl;

    double err = 0; double aux;
    for (int i = 0; i < n; ++i) {
        aux = 0;
        for (int j = 0; j < n; ++j) {
            aux += M[i*n+j]*x[j];
        }
        err += (b[i]-aux)*(b[i]-aux);
    }
    cout << "Error : " << err << endl;
}

void testLU() {
    int n = 3;
    double M[n*n], b[n], x[n];
    int P[n];

    // Input
    M[0] = 2.0; M[1] = -1.0; M[2] = 0.0;
    M[3] = -1.0; M[4] = 2.0; M[5] = -1.0;
    M[6] = 0.0; M[7] = -1.0; M[8] = 2.0;

    b[0] = 1.0; b[1] = 2.0; b[2] = 3.0;


    // Solve system
    LU(n,M,P);
    solveLU(n, M , P , b, x);


    // Output
    cout << "L = " << endl;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cout << M[i*n + j] << " , ";
        }
        cout << endl;
    }
    cout << endl;

    cout << "P = " << endl;
    for (int i = 0; i < n; ++i) {
        cout << P[i] << " , ";
    }
    cout << endl;

    cout << "x = " << endl;
    for (int i = 0; i < n; ++i) {
        cout << x[i] << " , ";
    }
    cout << endl;
}

void testCG() {
    int n = 3;
    double M[n*n], b[n], x[n], x0[n], r[n];

    // Input
    M[0] = 2.0; M[1] = -1.0; M[2] = 0.0;
    M[3] = -1.0; M[4] = 2.0; M[5] = -1.0;
    M[6] = 0.0; M[7] = -1.0; M[8] = 2.0;

    b[0] = 1.0; b[1] = 2.0; b[2] = 3.0;

    x0[0] = 1.0; x0[1] = 1.0; x0[2] = 1.0;


    // Solve system
    conjugateGradient CG(n, M, b);
    CG.configure(0.000001, false, 1);
    CG.solve(x0);

    CG.getSolution(x);
    CG.getResidual(r);

    double err = 0; double aux;
    for (int i = 0; i < n; ++i) {
        aux = 0;
        for (int j = 0; j < n; ++j) {
            aux += M[i*n+j]*x[j];
        }
        err += (b[i]-aux)*(b[i]-aux);
    }
    cout << "Error1 : " << err << endl;

    // Output
    cout << "Converged : " << CG.getConvergence() << endl;
    cout << "Number of iterations: " << CG.getNumIter() << endl;
    cout << "Error2 : " << CG.getError() << endl;

    cout << "x = " << endl;
    for (int i = 0; i < n; ++i) {
        cout << x[i] << " , ";
    }
    cout << endl;
}
